private/pkg/git/gitbranch/gitbranch.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitbranch
+
+import (
+	"errors"
+
+	"github.com/bufbuild/buf/private/pkg/git/gitobject"
+)
+
+// Ranger ranges over branches and commits for a git repository.
+//
+// Only branches and commits from the remote named `origin` are ranged.
+type Ranger interface {
+	// BaseBranch is the base branch of the repository. This is either
+	// configured via the `WithBaseBranch` option, or discovered via the
+	// remote named `origin`. Therefore, discovery requires that the repository
+	// is pushed to the remote.
+	BaseBranch() string
+	// Branches ranges over branches in the repository in an undefined order.
+	Branches(func(branch string) error) error
+	// Commits ranges over commits for the target branch in reverse topological order.
+	//
+	// Parents are visited before children, and only left parents are visited (i.e.,
+	// commits from branches merged into the target branch are not visited).
+	Commits(branch string, f func(commit gitobject.Commit) error) error
```

Comment:
`ForEachCommit`

Do we want to make `branch` an option, does it ever make sense to range over every commit in a git repo?

Reply:
I don't think so, you want to start from some commit and travel the graph, this is how [`git-rev-list`](https://git-scm.com/docs/git-rev-list) works. For our purposes, passing an arbitrary commit is clunkier than passing a branch reference.